File: frontend/js/views/reporting_time_pay.js
// Reporting-time pay generator — clamped half-shift minimum guarantee and
// additional owed, via /calc/reporting-time-pay.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const hrs = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderReportingTimePay(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.rtp.h1.title">// REPORTING-TIME PAY</span></h1>
        <p class="muted small" data-i18n="view.rtp.hint.intro">
            When an employee reports for a scheduled shift but is sent home early, California law guarantees
            pay for half the scheduled shift, with a two-hour minimum and a four-hour maximum, at the regular
            rate. It computes the reporting-time minimum, the hours guaranteed, and any additional pay owed
            beyond the hours actually worked. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.rtp.h2.inputs">Shift inputs</h2>
            <form id="rtp-form" class="inline-form">
                <label><span data-i18n="view.rtp.label.state">State</span>
                    <input type="text" name="state" value="California" required></label>
                <label><span data-i18n="view.rtp.label.employer">Employer</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.rtp.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.rtp.label.scheduled">Scheduled hours</span>
                    <input type="number" step="0.5" min="0" name="scheduled_hours" value="8" required></label>
                <label><span data-i18n="view.rtp.label.worked">Hours worked</span>
                    <input type="number" step="0.25" min="0" name="hours_worked" value="1"></label>
                <label><span data-i18n="view.rtp.label.rate">Regular rate ($/hr)</span>
                    <input type="number" step="0.01" min="0" name="regular_rate_usd" value="20" required></label>
                <label><span data-i18n="view.rtp.label.min">Minimum hours</span>
                    <input type="number" step="0.5" min="0" name="min_hours" value="2"></label>
                <label><span data-i18n="view.rtp.label.max">Maximum hours</span>
                    <input type="number" step="0.5" min="0" name="max_hours" value="4"></label>
                <label><span data-i18n="view.rtp.label.shiftdate">Shift date</span>
                    <input type="date" name="shift_date" value="2026-06-20" required></label>
                <label><span data-i18n="view.rtp.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.rtp.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.rtp.ph.statute'))}"></label>
            </form>
        </div>
        <div id="rtp-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#rtp-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            employer_name: (fd.get('employer_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            scheduled_hours: Number(fd.get('scheduled_hours')) || 0,
            hours_worked: Number(fd.get('hours_worked')) || 0,
            regular_rate_usd: Number(fd.get('regular_rate_usd')) || 0,
            min_hours: Number(fd.get('min_hours')) || 0,
            max_hours: Number(fd.get('max_hours')) || 0,
            shift_date: fd.get('shift_date'),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcReportingTimePay(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.rtp.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const el = mount.querySelector('#rtp-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.rtp.card.guaranteed">Guaranteed pay</div>
                    <div class="value">${money(doc.guaranteed_pay_usd)}</div></div>
                <div class="card neg"><div class="label" data-i18n="view.rtp.card.additional">Additional owed</div>
                    <div class="value">${money(doc.additional_owed_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.rtp.card.min">Reporting minimum</div>
                    <div class="value">${hrs(doc.reporting_min_hours)} h</div></div>
                <div class="card"><div class="label" data-i18n="view.rtp.card.hours">Guaranteed hours</div>
                    <div class="value">${hrs(doc.guaranteed_hours)} h</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="rtp-copy" type="button" data-i18n="view.rtp.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="rtp-download" type="button" data-i18n="view.rtp.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#rtp-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.rtp.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.rtp.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#rtp-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'reporting-time-pay.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/reporting_time_pay.rs
//! Reporting-time pay — when an employee reports for a scheduled shift but is sent
//! home early (or given less work than scheduled), California law guarantees pay
//! for half the scheduled shift, with a two-hour minimum and a four-hour maximum,
//! at the regular rate. This computes the reporting-time minimum hours, the hours
//! guaranteed, and any additional pay owed beyond the hours actually worked. No
//! existing generator computes a clamped half-shift minimum guarantee. Drafting
//! aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct ReportingTimeInput {
    pub employer_name: String,
    pub employee_name: String,
    /// Hours the employee was scheduled to work.
    pub scheduled_hours: f64,
    /// Hours the employee actually worked before being sent home.
    #[serde(default)]
    pub hours_worked: f64,
    /// Regular rate of pay, per hour.
    pub regular_rate_usd: f64,
    /// Minimum reporting-time hours (California is 2).
    #[serde(default = "default_min_hours")]
    pub min_hours: f64,
    /// Maximum reporting-time hours (California is 4).
    #[serde(default = "default_max_hours")]
    pub max_hours: f64,
    pub shift_date: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

fn default_min_hours() -> f64 {
    2.0
}

fn default_max_hours() -> f64 {
    4.0
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ReportingTimePay {
    pub title: String,
    /// Half the scheduled shift, clamped to [min, max].
    pub reporting_min_hours: f64,
    /// The greater of hours worked and the reporting-time minimum.
    pub guaranteed_hours: f64,
    /// Pay for the guaranteed hours.
    pub guaranteed_pay_usd: f64,
    /// Additional pay owed beyond the hours actually worked.
    pub additional_owed_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &ReportingTimeInput) -> ReportingTimePay {
    let half = i.scheduled_hours / 2.0;
    // Clamp half the shift to the [min, max] reporting-time window.
    let reporting_min = half.max(i.min_hours).min(i.max_hours);
    let guaranteed_hours = i.hours_worked.max(reporting_min);
    let guaranteed_pay = cents(guaranteed_hours * i.regular_rate_usd);
    let additional_hours = (reporting_min - i.hours_worked).max(0.0);
    let additional = cents(additional_hours * i.regular_rate_usd);

    let calc_body = format!(
        "Half of the {}-hour scheduled shift is {} hours, which falls within the {}–{} hour reporting-time window, so the reporting-time minimum is {} hours. The employee worked {} hours; the greater of the two is guaranteed, for {} hours of pay, or {} at {} per hour. Additional pay owed beyond hours worked is {}.",
        cents(i.scheduled_hours),
        cents(half),
        cents(i.min_hours),
        cents(i.max_hours),
        cents(reporting_min),
        cents(i.hours_worked),
        cents(guaranteed_hours),
        money(guaranteed_pay),
        money(i.regular_rate_usd),
        money(additional)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("Reporting-time pay is owed under the wage laws of the State of {}.", i.state)
    } else {
        format!("Reporting-time pay is owed under the wage laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Employer: {}\nEmployee: {}\nShift date: {}\nStatement date: {}",
                i.employer_name, i.employee_name, i.shift_date, i.date
            ),
        },
        DocClause {
            heading: "1. Reporting-Time Pay".into(),
            body: "An employee who reports for a scheduled shift but is furnished less than half the scheduled hours must be paid for half the shift, with a two-hour minimum and a four-hour maximum, at the regular rate.".into(),
        },
        DocClause { heading: "2. Calculation".into(), body: calc_body },
        DocClause {
            heading: "3. Payment".into(),
            body: format!(
                "The employee is guaranteed {} for the shift. Of that, {} is additional reporting-time pay beyond the hours actually worked.",
                money(guaranteed_pay), money(additional)
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Certification".into(),
            body: format!(
                "Employer: ____________________  Date: __________\n{}",
                i.employer_name
            ),
        },
    ];

    ReportingTimePay {
        title: "Reporting-Time Pay Statement".into(),
        reporting_min_hours: cents(reporting_min),
        guaranteed_hours: cents(guaranteed_hours),
        guaranteed_pay_usd: guaranteed_pay,
        additional_owed_usd: additional,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> ReportingTimeInput {
        ReportingTimeInput {
            employer_name: "Acme Co".into(),
            employee_name: "Pat Worker".into(),
            scheduled_hours: 8.0,
            hours_worked: 1.0,
            regular_rate_usd: 20.0,
            min_hours: 2.0,
            max_hours: 4.0,
            shift_date: "2026-06-20".into(),
            date: "2026-07-01".into(),
            state: "California".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn half_shift_capped_at_four() {
        let d = generate(&base());
        // half of 8 = 4 → min 4 hours; worked 1 → guaranteed 4.
        assert!(close(d.reporting_min_hours, 4.0));
        assert!(close(d.guaranteed_hours, 4.0));
        assert!(close(d.guaranteed_pay_usd, 80.0));
        assert!(close(d.additional_owed_usd, 60.0));
    }

    #[test]
    fn short_shift_floored_at_two() {
        let d = generate(&ReportingTimeInput { scheduled_hours: 3.0, hours_worked: 0.5, ..base() });
        // half of 3 = 1.5 → floored to 2.
        assert!(close(d.reporting_min_hours, 2.0));
        assert!(close(d.additional_owed_usd, 30.0));
    }

    #[test]
    fn long_shift_capped_at_four() {
        let d = generate(&ReportingTimeInput { scheduled_hours: 10.0, hours_worked: 1.0, ..base() });
        // half of 10 = 5 → capped to 4.
        assert!(close(d.reporting_min_hours, 4.0));
        assert!(close(d.additional_owed_usd, 60.0));
    }

    #[test]
    fn worked_beyond_minimum_no_additional() {
        let d = generate(&ReportingTimeInput { hours_worked: 5.0, ..base() });
        assert!(close(d.additional_owed_usd, 0.0));
        assert!(close(d.guaranteed_hours, 5.0));
        assert!(close(d.guaranteed_pay_usd, 100.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&ReportingTimeInput { statute_citation: "IWC Wage Order 5".into(), ..base() });
        assert_eq!(d.statutory_citation, "IWC Wage Order 5");
        assert!(d.clauses.iter().any(|c| c.body.contains("IWC Wage Order 5")));
    }
}
